File: C++/Greedy/21_page_faults_in_LRU.cpp
/*
Idea:
using list/vector to simulate cache. iterating the array and counting total page faults with inserting new pages.
*/

//{ Driver Code Starts
// Initial Template for C++



#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution {
public:
    int pageFaults(int N, int C, int pages[])
    {
        int pageFaultCount = 0;  // Counter for page faults
        vector<int> cache;  // Vector to simulate the cache

        for (int i = 0; i < N; i++)
        {
            // Check if the page is present in the cache
            auto cacheIterator = find(cache.begin(), cache.end(), pages[i]);

            // Page fault
            if (cacheIterator == cache.end())
            {
                pageFaultCount++;

                // If the cache is full, remove the page at the front
                if (cache.size() == C)
                    cache.erase(cache.begin());

                // Add the new page to the end of the cache
                cache.push_back(pages[i]);
            }
            else
            {
                // If the page is present, move it to the end of the cache
                cache.erase(cacheIterator);
                cache.push_back(pages[i]);
            }
        }

        return pageFaultCount;  // Return the total number of page faults
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N, C;
        cin >> N;
        int pages[N];
        for (int i = 0;i < N;i++)
            cin >> pages[i];
        cin >> C;

        Solution ob;
        cout << ob.pageFaults(N, C, pages) << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Python/Greedy/21_page_faults_in_LRU.py
"""
Idea:
using list/vector to simulate cache. iterating the array and counting total page faults with inserting new pages.
"""


# User function Template for python3

class Solution:
    def pageFaults(self, N, C, pages):
        page_fault_count = 0  # Counter for page faults
        cache = []  # List to simulate the cache

        for i in range(N):
            # Check if the page is present in the cache
            page_index = next((index for index, page in enumerate(cache) if page == pages[i]), None)

            # Page fault
            if page_index is None:
                page_fault_count += 1

                # If the cache is full, remove the page at the front
                if len(cache) == C:
                    cache.pop(0)

                # Add the new page to the end of the cache
                cache.append(pages[i])
            else:
                # If the page is present, move it to the end of the cache
                cache.pop(page_index)
                cache.append(pages[i])

        return page_fault_count  # Return the total number of page faults


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        N = int(input())
        pages = input().split()
        for itr in range(N):
            pages[itr] = int(pages[itr])
        C = int(input())

        ob = Solution()
        print(ob.pageFaults(N, C, pages))

# } Driver Code Ends
